<launch>

  <!-- Set use_sim_time true for datasets-->
  <param name="use_sim_time" value="true" />

  <!-- static_transform_publisher x y z yaw pitch roll frame_id child_frame_id -->
  <!-- This transform was used set only to test if everything works
       when using something different than the camera as a reference frame.
       If enabled, take care to also fix the transform below. -->
  <!--node pkg="tf2_ros" type="static_transform_publisher" name="camera_broadcaster" args="1 0 0.5 -1.57 0 -1.57 base_link left_camera" /-->

  <!-- 'start' frame is the starting pose of the camera
       When 'base_link' is set as the reference frame,
       take care to set this transformation to the same as the previous one -->
  <node pkg="tf2_ros" type="static_transform_publisher" name="start_broadcaster" args="0 0 0 0 0 0 start map" />

  <!-- S-PTAM pose path publisher. Used for visualization. -->
  <node pkg="ros_utils" type="pose_to_path" name="sptam_path">
    <remap from="pose" to="sptam/robot/pose" />
    <remap from="path" to="sptam/robot/path" />
  </node>

  <!--
    Master nodelet manager.
    S-PTAM as well as the extraction nodelets will be run
    under a single manager to avoid unnecesary serialization of messages.
  -->
  <node name="nodelet_manager" pkg="nodelet" type="nodelet" args="manager" output="screen" clear_params="true" />

  <!-- S-PTAM node -->
  <node pkg="nodelet" type="nodelet" name="sptam" args="load sptam/sptam_nodelet nodelet_manager" output="screen" clear_params="true">

    <!-- Read S-PTAM parameters file -->
    <rosparam command="load" file="$(find sptam)/configurationFiles/kitti.yaml" />

    <param name="use_prediction" value="false" />
    <param name="approximate_sync" value="false" />
    <param name="publish_transform" value="true" />
    <param name="publish_on_fail" value="true" />

    <!--param name="prediction_frame" value="odom" /-->
    <param name="base_frame" value="left_camera" />
    <param name="camera_frame" value="left_camera" />
    <!--param name="map_frame" value="map" /-->
    <param name="reference_frame" value="left_camera" />

    <remap from="/stereo/left/image_rect" to="/kitti_stereo/left/image_rect" />
    <remap from="/stereo/right/image_rect" to="/kitti_stereo/right/image_rect" />
    <remap from="/stereo/left/camera_info"  to="/kitti_stereo/left/camera_info" />
    <remap from="/stereo/right/camera_info"  to="/kitti_stereo/right/camera_info" />

  </node>

  <!-- DENSE node -->
  <node pkg="nodelet" type="nodelet" name="dense" args="load dense/denseNodelet nodelet_manager" output="screen" clear_params="true">

    <param name="approximate_sync" value="false" />
    <param name="base_frame" value="left_camera" />
    <param name="camera_frame" value="left_camera" />

    <param name="disp_calc_method" value="libelas"/>
    <param name="max_distance" value="20"/>
    <param name="VoxelLeafSize" value="0.1"/>

    <param name="stereoscan_threshold" value="0.25"/>
    <!--param name="fusion_heuristic" value="simpleMean"/-->
    <param name="fusion_heuristic" value="weigthDistances"/>

    <param name="FrustumNearPlaneDist" value="0.0001"/>
    <param name="FrustumFarPlaneDist" value="50"/>
    <param name="libelas_ipol_gap" value="1000"/>

    <param name="local_area_size" value="10"/>
    <param name="pub_area_filter_min" value="0"/>

    <param name="refinement_linear_threshold" value="0.01"/>
    <param name="refinement_angular_threshold" value="0.001"/>

    <remap from="/dense/keyframes" to="/sptam/keyframes"/>

  </node>
  <!--
    <param name="single_depth_map_clouds" value="/BACKUP/tesina/datasets/kitti/output/"/>
    <param name="single_depth_map_poses" value="/BACKUP/tesina/datasets/kitti/output/poses.txt"/>
    <param name="single_depth_map_mode" value="kitti_global"/>
    <param name="single_depth_map_region_size" value="30"/>

    <param name="single_depth_map_clouds" value="/BACKUP/tesina/datasets/kitti/sequences/00/velodyne/pcd/"/>
    <param name="single_depth_map_poses" value="/BACKUP/tesina/datasets/kitti/sequences/00/velodyne/poses.txt"/>
    <param name="single_depth_map_mode" value="kitti_global"/>
    <param name="single_depth_map_region_size" value="30"/>

    <param name="single_cloud_path" value="/BACKUP/tesina/datasets/kitti/sequences/00/dense/pcd/"/>
  -->

</launch>
